### DISCUSSION TITLE
How does `make bootstrap` work?

### DISCUSSION THREAD

**Discussion by unfode** [ocaml/dune#8239] [General]
When I run `make bootstrap`, the first thing that is executed seems to be `ocamlc ...`, but I don't see it in the `Makefile`.

![image](https://github.com/ocaml/dune/assets/95689995/ad5fcbb2-15b5-44cd-973d-93a1e98aaac8)

**Comment by Alizter** [comment]
This recently changed. The target you are looking for in the makefile is $(BIN). Which is _boot/dune.exe.

**Reply by unfode** [reply]
I don't know much about Makefiles. Do you mean that the `bootstrap` target delegates to the `$(BIN)` target?
![image](https://github.com/ocaml/dune/assets/95689995/7c7dc904-da9f-45a6-8efa-e225ba2ec767)

**Reply by Alizter** [reply]
What you are really doing when you do `make bootstrap` is recalling the makefile in `-B` mode (which will always rebuild for reasons I'll explain later). The target `_boot/dune.exe` is the bootstrapped version of Dune which is put together by running the script `boot/bootstrap.ml`.

The reason we pass `-B` is so that you can run `make bootstrap` anytime you like without `make` rejecting the request since `_boot/dune.exe` will be "up-to-date". You don't need to bootstrap manually in many cases since the other targets in the makefile will ask for the bootstrapped version and then proceed to use it to build whatever.

**Reply by Alizter** [reply]
The `@` in front of `ocaml` in the makefile hides the output. If you remove the `@` from the _boot/dune.exe target you will see the order of execution. First the `bootstrap.ml` script it run, then `ocamlc` within that script and then finally `_boot/dune.exe` is built. You can run it from `./dune.exe` which points to `_boot/dune.exe`.

**Reply by unfode** [reply]
So
```makefile
bootstrap:
	$(MAKE) -B _boot/dune.exe
```
is equivalent to
```makefile
bootstrap:
	$(MAKE) -B $(BIN)
```
?

Why not use the latter? It's clearer. The former causes a misconception in me that `_boot/dune.exe` is a file rather than a target.

**Reply by Alizter** [reply]
Good point, this was changed earlier this week. I simply copied the command I did for manually bootstrapping.

**Reply by unfode** [reply]
Thank you! I'll create a pull request for this.
